File: 프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.java
import java.util.*;
class Solution {
    public int solution(int[][] board, int[] moves) {
        int answer = 0;
        Stack<Integer> stack = new Stack<>();
        for(int i = 0; i < board.length; i++) {
            if(board[i][moves[0]-1] != 0) {
                stack.push(board[i][moves[0]-1]);
                board[i][moves[0]-1] = 0;
                break;
            }
        }
        for(int i = 1; i < moves.length; i++) {
            for(int j = 0; j < board.length; j++) {
                if(board[j][moves[i]-1] != 0) {
                    if(!stack.isEmpty() && stack.peek() == board[j][moves[i]-1]) {
                        System.out.print(board[j][moves[i]-1]);
                        stack.pop();
                        answer++;
                    }
                    else {
                        stack.push(board[j][moves[i]-1]);    
                    }
                    board[j][moves[i]-1] = 0;
                    break;
                }
            }
        }
        return answer*2;
    }
}

File: 프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.js
function solution(board, moves) {
    var answer = 0;
    const lanes = Array.from({length:board.length}, ()=> []);
    
    for (let i = board.length - 1; i >= 0; i--) {
        for (let j = 0; j < board[i].length; j++) {
            if (board[i][j]) lanes[j].push(board[i][j]);
        }
    }
    
    const stack = [];
    
    for (const move of moves) {
        if (lanes[move-1].length > 0) { 
            const doll = lanes[move-1].pop();
            if (stack.length > 0 && stack[stack.length-1] === doll) {
                answer+=2;
                stack.pop();
            } else stack.push(doll);
            
        }
    }
    return answer;
}